crates/proof-of-sql/src/sql/proof_exprs/range_check.rs

```diff
@@ -23,44 +116,315 @@ fn decompose_scalar_to_words<'a, S: Scalar + 'a>(
+#[allow(clippy::missing_panics_doc)]
+#[allow(clippy::cast_possible_truncation)]
+fn prove_word_values<'a, S: Scalar + 'a>(
+    alloc: &'a Bump,
+    scalars: &[S],
+    alpha: S,
+    table_length: usize,
+    builder: &mut FinalRoundBuilder<'a, S>,
+) {
+    // Allocate from 0 to 255 and pertrub with verifier challenge
+    let word_values_plus_alpha: &mut [S] =
+        alloc.alloc_slice_fill_with(max(256, scalars.len()), |i| S::from(&(i as u8)) + alpha);
+    builder.produce_intermediate_mle(word_values_plus_alpha as &[_]);
+
+    // Now produce an intermediate MLE over the inverted word values + verifier challenge alpha
+    let inverted_word_values_plus_alpha: &mut [S] = alloc.alloc_slice_fill_with(256, |i| {
+        S::try_from(i.into()).expect("word value will always fit into S") + alpha
+    });
+    slice_ops::batch_inversion(&mut inverted_word_values_plus_alpha[..]);
+    builder.produce_intermediate_mle(inverted_word_values_plus_alpha as &[_]);
+
+    let input_ones = alloc.alloc_slice_fill_copy(table_length, true);
+
+    // Argument:
+    // (word_values + α)⁻¹ * (word_values + α) - 1 = 0
+    builder.produce_sumcheck_subpolynomial(
+        SumcheckSubpolynomialType::Identity,
+        vec![
+            (
+                S::one(),
+                vec![
+                    Box::new(inverted_word_values_plus_alpha as &[_]),
+                    Box::new(word_values_plus_alpha as &[_]),
+                ],
+            ),
+            (-S::one(), vec![Box::new(input_ones as &[_])]),
+        ],
+    );
+}
+
+/// Argue that the sum of all words in each row, minus the count of each word
+/// multiplied by the inverted word value, is zero.
+///
+/// ```text
+/// ∑ (I₀ + I₁ + I₂ + I₃ - (C * IN)) = 0
+/// ```
+///
+/// Where:
+/// - `I₀, I₁, I₂, I₃` are the inverted word columns.
+/// - `C` is the count of each word.
+/// - `IN` is the inverted word values column.
+#[allow(clippy::missing_panics_doc)]
+fn prove_row_zero_sum<'a, S: Scalar + 'a>(
+    builder: &mut FinalRoundBuilder<'a, S>,
+    word_counts: &'a mut [i64],
+    alloc: &'a Bump,
+    scalars: &[S],
+    inverted_word_columns: &[&mut [S]],
+    alpha: S,
+) {
+    // Produce an MLE over the counts of each word value
+    builder.produce_intermediate_mle(word_counts as &[_]);
+
+    // Allocate row_sums from the bump allocator, ensuring it lives as long as 'a
+    let row_sums = alloc.alloc_slice_fill_with(scalars.len(), |_| S::ZERO);
+
+    // Iterate over each column and sum up the corresponding row values
+    for column in inverted_word_columns {
+        // Iterate over each scalar in the column
+        for (i, inv_word) in column.iter().enumerate() {
+            row_sums[i] += *inv_word;
+        }
+    }
 
-        // Invert all the terms in the column at once
-        slice_ops::batch_inversion(&mut terms_to_invert);
+    // Pass the row_sums reference with the correct lifetime to the builder
+    builder.produce_intermediate_mle(row_sums as &[_]);
```

Comment:
We do not need to commit to this because we can derive the evaluation of the sum as the sum of the evaluations.

Reply:
Does this increase the cost of verification?

Reply:
We understand we are trying to satisfy the following:

```SUM (I0+I1+I2+I3-(C*IN))=0```

and that the verifier has access to

```SUM (I0+I1+I2+I3) ```

but it doesnt appear that we currently have access to the ```(C*IN)``` part, maybe the prover supplies it, but its unclear how they might do that